// @accessible/visually-hidden/dist/umd/visually-hidden.dev.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('react')) :
  typeof define === 'function' && define.amd ? define(['exports', 'react'], factory) :
  (global = global || self, factory(global.VisuallyHidden = {}, global.React));
}(this, (function (exports, React) { 'use strict';

  function _extends() {
    _extends = Object.assign || function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];

        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }

      return target;
    };

    return _extends.apply(this, arguments);
  }

  var VisuallyHidden = function VisuallyHidden(_ref) {
    var children = _ref.children;
    return /*#__PURE__*/React.cloneElement(children, _extends({}, children.props, {
      style: _extends({}, visuallyHidden, children.props.style)
    }));
  };
  var visuallyHidden = {
    border: 0,
    clip: 'rect(0 0 0 0)',
    height: 1,
    width: 1,
    margin: -1,
    padding: 0,
    overflow: 'hidden',
    position: 'absolute'
  };

  if (typeof process !== 'undefined' && "production" !== 'production') {
    VisuallyHidden.displayName = 'VisuallyHidden';
  }

  exports.VisuallyHidden = VisuallyHidden;
  exports.visuallyHidden = visuallyHidden;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
